pipeline-runner/R/gem2s-5-create_seurat.R

```diff
@@ -1,132 +1,129 @@
-#  - Create a seurat object per sample
-#  - Adding emtpyDrops,  scrublet and MT-content
-#  - Create a file with flag_filtered
-
-# LOADING SPARSE MATRIX AND CONFIGURATION
-create_seurat <- function(input, pipeline_config) {
-  message("reloading old matrices...")
-  scdata_list <- readRDS("/output/pre-doublet-scdata_list.rds")
-
-  message("Loading configuration...")
-  config <- RJSONIO::fromJSON("/input/meta.json")
-
-  print_config(5, "Create Seurat", input, pipeline_config, config)
-
-  # Check which samples have been selected. Otherwiser we are going to use all of them.
-  if (length(config$samples) > 0) {
-    samples <- config$samples
-  } else {
-    samples <- names(scdata_list)
+#' Create a SeuratObject per sample
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out  'output' slot from call to \code{score_doublets}
+#'
+#' @return \code{prev_out} with added slot 'scdata_list' containing \code{SeuratObject}'s for each sample.
+#'
+#' @export
+#'
+create_seurat <- function(input, pipeline_config, prev_out) {
+
+  message("Creating Seurat Objects...")
+  message("items in prev_out: ", paste(names(prev_out), collapse = ' - '))
+
+  # destructure previous output
+  list2env(prev_out, envir = environment())
+
+  samples <- names(counts_list)
+  scdata_list <- list()
+  for (sample in samples) {
+    message("\nCreating SeuratObject for sample --> ", sample)
+
+    scdata_list[[sample]] <- construct_scdata(
+      counts = counts_list[[sample]],
+      doublet_score = doublet_scores[[sample]],
+      edout = edrops[[sample]],
+      sample = sample,
+      annot = annot,
+      config = config)
```

Comment:
I am thinking here that all these came out of `prev_out` but it might be that my knowledge of R is not very great that I don't understand how exactly they were taken from there. Maybe the call of this function `list2env(prev_out, envir = environment())` on line 16 somehow magically loaded them in? @alexvpickering can you explain how they got here?

Reply:
Yes that is correct. It takes a list of named objects and add them to the environment specified. It's a more succinct version of this:

```R
  # destructure previous output
  config <- prev_out$config
  annot <- prev_out$annot
  edrops <- prev_out$edrops
  counts_list <- prev_out$counts_list
  doublet_scores <- prev_out$doublet_scores
```